File: OCH_Tao/tessoku-book/A10.js
const input = require("fs").readFileSync("/dev/stdin", "utf8").trim().split("\n");
const N = parseInt(input.shift(), 10);
const A = input.shift().split(" ").map(x => parseInt(x, 10));
let P = Array(100009).fill(0);
let Q = Array(100009).fill(0);
for (let i = 1; i < N + 1; i++) {
  P[i] = Math.max(P[i - 1], A[i - 1]);
}
for (let i = 1; i < N + 1; i++) {
  Q[i] = Math.max(Q[i - 1], A[N - i]);
}
input.shift();
for (const i of input) {
  const [L, R] = i.split(" ").map(x => parseInt(x, 10));
  console.log(Math.max(P[L - 1], Q[N - R]));
}

File: OCH_Tao/tessoku-book/A10.py
N = int(input())
A = list(map(int, input().split()))
P = [0]*(N+9)
Q = [0]*(N+9)
for i in range(N):
  P[i+1] = max(P[i], A[i])
for i in range(N):
  Q[i+1] = max(Q[i], A[-(i+1)])
D = int(input())
for i in range(D):
  L, R = map(int, input().split())
  print(max(P[L-1], Q[N-R]))
